File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge32.dart
/*
 * Reto #32
 * EL SEGUNDO
 * Fecha publicación enunciado: 08/08/22
 * Fecha publicación resolución: 15/08/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Dado un listado de números, encuentra el SEGUNDO más grande.
 *
 * Información adicional:
 * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la comunidad.
 * - Tienes toda la información sobre los retos semanales en https://retosdeprogramacion.com/semanales2022.
 *
 */

import 'dart:math'; //Para listas aleatoreas

num secondMax(List<num> list) {
  if (list.isEmpty) throw Exception("lista vacia");
  List<num> maxs = List<num>.filled(2, list.first);
  for (var n in list) {
    if (maxs.last < n) {
      maxs.first = maxs.last;
      maxs.last = n;
    } else if (maxs.first < n || maxs.first == maxs.last) maxs.first = n;
  }
  return maxs.first;
}

void main(List<String> args) {
  Random rand = Random();
  List<num> list =
      List<num>.generate(rand.nextInt(25), (index) => rand.nextInt(100));

  print(list);
  try {
    print(secondMax(list));
  } catch (e) {
    print(e);
  }
}

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge32.py
""" package com.mouredev.weeklychallenge2022

/*
 * Reto #32
 * EL SEGUNDO
 * Fecha publicación enunciado: 08/08/22
 * Fecha publicación resolución: 15/08/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Dado un listado de números, encuentra el SEGUNDO más grande.
 *
 * Información adicional:
 * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la comunidad.
 * - Tienes toda la información sobre los retos semanales en https://retosdeprogramacion.com/semanales2022.
 *
 */

fun main() {
    println(findSecondGreater(arrayListOf(4, 6, 1, 8, 2)))
    println(findSecondGreater(arrayListOf(4, 6, 8, 8, 6)))
    println(findSecondGreater(arrayListOf(4, 4)))
    println(findSecondGreater(arrayListOf()))
}

fun findSecondGreater(numbers: List<Int>): Int? {

    val sortedNumbers = mutableListOf<Int>()

    for (number in numbers) {

        var found = false

        for ((index, sortedNumber) in sortedNumbers.withIndex()) {

            if (number >= sortedNumber) {
                if (number != sortedNumber) {
                    sortedNumbers.add(index, number)
                }
                found = true
                break
            }
        }

        if (!found) {
            sortedNumbers.add(number)
        }
    }

    return if(sortedNumbers.count() >= 2) sortedNumbers[1] else null
} """


def second_greater(numbers: list):

    sorted_numbers = []

    for number in numbers:

        added = False

        for index, sorted_number in enumerate(sorted_numbers):

            if number < sorted_number:
                sorted_numbers.insert(index, number)
                added = True
                break

        if not added:
            sorted_numbers.append(number)

    return sorted_numbers[-2] if len(sorted_numbers) > 1 else None


print(second_greater([4, 6, 1, 8, 2]))
print(second_greater([]))